haystack/components/builders/dynamic_chat_prompt_builder.py

```diff
@@ -0,0 +1,273 @@
+    prompt_builder = DynamicChatPromptBuilder(runtime_variables=["documents"])
+    llm = GPTChatGenerator(api_key="<your-api-key>", model_name="gpt-3.5-turbo")
+
+
+    @component
+    class DocumentProducer:
+
+        @component.output_types(documents=List[Document])
+        def run(self, doc_input: str):
+            return {"documents": [Document(content=doc_input)]}
+
+
+
+    pipe = Pipeline()
+    pipe.add_component("doc_producer", DocumentProducer())
+    pipe.add_component("prompt_builder", prompt_builder)
+    pipe.add_component("llm", llm)
+
+    # note here how prompt_builder.documents is received from doc_producer.documents
+    pipe.connect("doc_producer.documents", "prompt_builder.documents")
+    pipe.connect("prompt_builder.prompt", "llm.messages")
+
+    messages = [ChatMessage.from_system("Be helpful assistant, but brief!"),
+                ChatMessage.from_user("Here is the document: {{documents[0].content}} Now, answer the
+                following: {{query}}")]
+
+
+    pipe.run(data={"doc_producer": {"doc_input": "Hello world, I'm Haystack!"},
+                   "prompt_builder": {"prompt_source": messages,
+                                      "template_variables":{"query": "who's making a greeting?"}}})
+
+    >> {'llm': {'replies': [ChatMessage(content='Haystack', role=<ChatRole.ASSISTANT: 'assistant'>, name=None,
+    >> metadata={'model': 'gpt-3.5-turbo-0613', 'index': 0, 'finish_reason': 'stop', 'usage':
+    >> {'prompt_tokens': 51, 'completion_tokens': 2, 'total_tokens': 53}})]}}
+    ```
+    """
+
+    def __init__(self, runtime_variables: Optional[List[str]] = None):
+        """
+        Initializes DynamicChatPromptBuilder with the provided variable names. These variable names are used to resolve
+        variables and their values during pipeline runtime execution. For example, if `runtime_variables` contains
+        `documents` your instance of DynamicChatPromptBuilder will expect an input called `documents`.
+        The values associated with variables from the pipeline runtime are then injected into template placeholders
+        of a ChatMessage that is provided to the `run` method. See `run` method for more details.
+
+        :param runtime_variables: A list of template variable names you can use in chat prompt construction.
+        :type runtime_variables: Optional[List[str]]
+        """
+        runtime_variables = runtime_variables or []
+
+        # setup inputs
+        run_input_slots = {"prompt_source": List[ChatMessage], "template_variables": Optional[Dict[str, Any]]}
+        kwargs_input_slots = {var: Optional[Any] for var in runtime_variables}
+        component.set_input_types(self, **run_input_slots, **kwargs_input_slots)
+
+        # setup outputs
+        component.set_output_types(self, prompt=List[ChatMessage])
+        self.runtime_variables = runtime_variables
+
+    def to_dict(self) -> Dict[str, Any]:
```

Comment:
This is just doing the same as the default implementation so we can remove the method.